wordpress-mobile/WordPress-iOS
# Reblog Action: Users with one available site

### This PR adds the reblog action to the app. It covers the use case of users with one available site.

Ref: #12960

### Description
The goal of this PR is to add the reblog action to the app, and connect it to the reblog buttons in either `ReaderPostCardCell` and `ReaderDetailViewController`.
At this moment, it only covers the use case of users who have only one available site. For this use case, when the `reblog` button is tapped, the block editor will open, and its editable content will consist of an image, a quote from the original post and a citation containing a link to the original post. Below is a short video that illustrates the behavior of the feature when tapping on the `reblog` button in the post detail, in the reader.


![Dec-13-2019 17-15-34](https://user-images.githubusercontent.com/34376330/70838195-bc6cc400-1dcc-11ea-82c3-055616f40730.gif)

This same behavior also happens when tapping on the `reblog` button in a cell of the post list (see sequence of screenshots below).

![Screen Shot 2019-12-13 at 6 08 35 PM](https://user-images.githubusercontent.com/34376330/70839714-a7dffa00-1dd3-11ea-9f1b-6f05616e6782.png)


Change log:

- Added a new `ReaderReblogAction.swift` file, containing 
  - a ReaderReblocAction type, with an execute(_ :, _ :) method, triggered by a UI action
  - a ReblogPresenter type, responsible of presenting the appropriate screens, depending on how many available sites are there
  - a ReblogFormatter type, responsible of formatting the content in a Gutenberg-ready html, which will trigger the block editor
  - an extension to `Post' that adds a method to prepare the content for reblogging
- Added @IBAction methods connected to the reblog buttons, in `ReaderPostCardCell.swift` and `ReaderDetailViewController.swift`, which will trigger ReaderReblogAction.execute(_:, _:)
- Added a new method `func readerCell(_ cell: ReaderPostCardCell, reblogActionForProvider provider: ReaderPostContentProvider)` in the `ReaderPostCardCellDelegate` protocol
- Added unit tests for the new objects

#### Note
1. The use cases of 
  - users with zero available sites
  - users with multiple (more than 1) available sites
are yet to come, and not included in this PR.

2. Analytics (tracks) are not yet included.

### To test:
- Make sure you have credentials for a test user that has only one available site.
  - if your test user has zero available  sites, or more than one available site, tapping on the `reblog` buttons will have no effect
- Update `wpcom_app_credentials` as needed.
- Build and run the app, and log in
- Select the reader in the tab bar
- From the `STREAMS` section, select the followed site(s) of your preference
- In the post list view, tap the `reblog` button on any post that has reblog enabled
- check that the content of the editor matches the description above
- tap on the ellipsis, and test that all the functionalities are enabled:
  - save as draft
  - switch to classic editor
  - switch to HTML mode
  - preview
  - post settings
- check that the post can be edited in both the block and the classic editor
- check that the label in the upper left matches the available site of your test user
- tap `Publish`, and check that that an action sheet containing the actions `Publish` and `Keep Editing` appears at the bottom of the screen
- Choose the `Publish` action
- Check that a `Post Published` toast, with the name title of the post and a `view` action appears at the bottom
- Tap that `view` action and confirm that it opens the viewer with the post
- Navigate to your site (either from the reader on my sites) and confirm that the post is present (pulldown to refresh might be needed)

PR submission checklist:

   - Unit tests have been added.

   - Since it's still a WIP feature, it shouldn't be necessary.
